README.md
# RedisSessionHandler

[![Pipeline Status](https://gitlab.com/1ma/RedisSessionHandler/badges/master/pipeline.svg)](https://gitlab.com/1ma/RedisSessionHandler/pipelines) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/1ma/RedisSessionHandler/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/1ma/RedisSessionHandler/?branch=master) [![Latest Stable Version](https://poser.pugx.org/uma/redis-session-handler/v/stable)](https://packagist.org/packages/uma/redis-session-handler) [![Monthly Downloads](https://poser.pugx.org/uma/redis-session-handler/d/monthly)](https://packagist.org/packages/uma/redis-session-handler)

An alternative Redis session handler featuring session locking and session fixation protection.


## News

* phpredis v4.1.0 (released on 2018-07-10) added support for session locking, but it is disabled by default. To enable
  it you must set the new `redis.session.locking_enabled` INI directive to `true`. This version is the first to pass
  the test in `ConcurrentTest` that stresses the locking mechanism.


## Installation

RedisSessionHandler requires PHP >=5.6 with the phpredis extension enabled and a Redis >=2.6 endpoint. Add [`uma/redis-session-handler`](https://packagist.org/packages/uma/redis-session-handler) to the `composer.json` file:

    $ composer require uma/redis-session-handler

Overwrite the default session handler with `UMA\RedisSessionHandler` before your application calls
any `session_` function. If you are using a framework and unsure when or where that happens, a
good rule of thumb is "as early as possible". A safe bet might be the frontend controller in the
public directory of the project or an equivalent initialization script.

```php
// top of my-project/web/app.php

require_once __DIR__ . '/../vendor/autoload.php';

session_set_save_handler(new \UMA\RedisSessionHandler(), true);
```

Note that calling `session_set_save_handler` overwrites any value you might have set in the `session.save_handler` option
of the php.ini file, so you don't need to change that. However, RedisSessionHandler still uses `session.save_path` to find
the Redis server, just like the vanilla phpredis session handler:

```ini
; examples
session.save_path = "localhost"
session.save_path = "localhost?timeout=2.5"
session.save_path = "tcp://1.2.3.4:5678?prefix=APP_SESSIONS:&database=2"
session.save_path = "unix:///var/run/redis.sock"
session.save_path = "/var/run/redis.sock?database=2"
```

Available query params:

* `timeout` (float), default `0.0`, which means unlimited timeout
* `prefix` (string), default `'PHPREDIS_SESSION:'`
* `auth` (string), default `null`
* `database` (int), default `0`

Currently only a single host definition is supported.

Note than when connecting through a Unix socket the timeout is ignored.


## Known Caveats

### Using RedisSessionHandler with the `max_execution_time` directive set to `0` is not recommended

Whenever it can, the handler uses the `max_execution_time` directive as a hard timeout for the session lock. This is a
last resort mechanism to release the session lock even if the PHP process crashes and the handler fails to do it itself.

When `max_execution_time` is set to `0` (meaning there is no maximum execution time) this kind of hard timeout cannot be used, as the lock
must be kept for as long as it takes to run the script, which is an unknown amount of time. This means that if for some unexpected reason
the PHP process crashes and the handler fails to release the lock there would be no safety net and you'd end up with a dangling lock
that you'd have to detect and purge by other means.

So when using RedisSessionHandler it is advised _not_ to disable `max_execution_time`.


### RedisSessionHandler does not support `session.use_trans_sid=1` nor `session.use_cookies=0`

When these directives are set this way PHP switches from using cookies to passing the session ID around as a query param.

RedisSessionHandler cannot work in this mode. _This is by design_.


### RedisSessionHandler ignores the `session.use_strict_mode` directive

Because running PHP with strict mode disabled (which is the default!) does not make any sense whatsoever. RedisSessionHandler
only works in strict mode. The _Session fixation_ section of this README explains what that means.


## Motivation

The Redis session handler bundled with [phpredis](https://github.com/phpredis/phpredis) has had a couple of rather serious
bugs for years, namely the [lack of per-session locking](https://github.com/phpredis/phpredis/issues/37) and the [impossibility to protect against session fixation attacks](https://github.com/phpredis/phpredis/issues/1033).

This package provides a compatible session handler built on top of the Redis extension that is not affected by these issues.


### Session Locking explained

In the context of PHP, "session locking" means that when multiple requests with the same session ID hit the server roughly
at the same time, only one gets to run while the others get stuck waiting inside `session_start()`. Only when that first request
finishes or explicitly runs [`session_write_close()`](http://php.net/manual/en/function.session-write-close.php), one of the others can move on.

When a session handler does not implement session locking concurrency bugs might start to surface under
heavy traffic. I'll demonstrate the problem using the default phpredis handler and this simple script:

```php
<?php

// a script that returns the total number of
// requests made during a given session's lifecycle.

session_start();

if (!isset($_SESSION['visits'])) {
  $_SESSION['visits'] = 0;
}

$_SESSION['visits']++;

echo $_SESSION['visits'];
```

First, we send a single request that will setup a new session. Then we use the session ID returned in
the `Set-Cookie` header to send a burst of 200 concurrent, authenticated requests.

```bash
$ http localhost/visit-counter.php
HTTP/1.1 200 OK
Cache-Control: no-store, no-cache, must-revalidate
Connection: keep-alive
Content-Type: text/html; charset=UTF-8
Date: Mon, 23 Jan 2017 12:30:17 GMT
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Pragma: no-cache
Server: nginx/1.11.8
Set-Cookie: PHPSESSID=9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21; path=/; HttpOnly
Transfer-Encoding: chunked

1

$ hey -n 200 -H "Cookie: PHPSESSID=9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21;" http://localhost/visit-counter.php
All requests done.

Summary:
  Total:	0.4033 secs
  Slowest:	0.1737 secs
  Fastest:	0.0086 secs
  Average:	0.0805 secs
  Requests/sec:	495.8509

Status code distribution:
  [200]	200 responses
```

Everything looks fine from the outside, we got the expected two hundred OK responses, but if we peek inside the Redis
database we see that the counter is way off. Instead of 201 visits we see a random number that is way lower than that:

```
127.0.0.1:6379> KEYS *
1) "PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21"

127.0.0.1:6379> GET PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21
"visits|i:134;"
```

Looking at Redis' `MONITOR` output we can see that under heavy load, Redis often executes two or more `GET` commands
one after the other, thus returning the same number of visits to two or more different requests. When that happens, all
those unlucky requests pass the same number of visits back to Redis, so some of them are ultimately lost. For instance, in this excerpt
of the log you can see how the 130th request is not accounted for.

```
1485174643.241711 [0 172.21.0.2:49780] "GET" "PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21"
1485174643.241891 [0 172.21.0.2:49782] "GET" "PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21"
1485174643.242444 [0 172.21.0.2:49782] "SETEX" "PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21" "900" "visits|i:129;"
1485174643.242878 [0 172.21.0.2:49780] "SETEX" "PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21" "900" "visits|i:129;"
1485174643.244780 [0 172.21.0.2:49784] "GET" "PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21"
1485174643.245385 [0 172.21.0.2:49784] "SETEX" "PHPREDIS_SESSION:9mcjmlsh9gp0conq7i5rci7is8gfn6s0gh8r3eub3qpac09gnh21" "900" "visits|i:130;"
```

RedisSessionHandler solves this problem with a "lock" entry for every session that only one thread of execution can create at a time.


### Session fixation explained

[Session fixation](https://www.owasp.org/index.php/Session_fixation) is the ability to choose your own session ID as an HTTP client. When clients are allowed to choose their
session IDs, a malicious attacker might be able to trick other clients into using an ID already known to him, then wait for them log in and hijack their session.

Starting from PHP 5.5.2, there's an INI directive called [`session.use_strict_mode`](http://php.net/manual/en/session.configuration.php#ini.session.use-strict-mode) to protect
PHP applications against such attacks. When "strict mode" is enabled and a unknown session ID is received, PHP should ignore it and generate a new
one, just as if it was not received at all. Unfortunately the phpredis handler ignores that directive and always trust whatever session ID is received from
the HTTP request.

```bash
$ http -v http://localhost/visit-counter.php Cookie:PHPSESSID=madeupkey
GET / HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Cookie: PHPSESSID=madeupkey     <----
Host: 127.0.0.1
User-Agent: HTTPie/0.9.6

HTTP/1.1 200 OK
Cache-Control: no-store, no-cache, must-revalidate
Connection: close
Content-type: text/html; charset=UTF-8
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Host: 127.0.0.1
Pragma: no-cache

1

$ redis-cli

127.0.0.1:6379> keys *
1) "PHPREDIS_SESSION:madeupkey"

127.0.0.1:6379> GET PHPREDIS_SESSION:madeupkey
"visits|i:1;"
```

Hence RedisSessionHandler only works in strict mode. It only accepts external session IDs
that are already inside the Redis store.


## Testing

### Running the tests

To do that you'll need Docker >=1.10 and docker-compose >=1.8.

In order to run the integration test suite just type `composer test` and it will take care of installing
the dev dependencies, setting up the testing containers and running the tests.

```bash
$ composer test
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Generating autoload files
> docker-compose -f tests/docker-compose.yml up -d
tests_redis_1 is up-to-date
tests_fpm56_1 is up-to-date
tests_fpm71_1 is up-to-date
tests_fpm70_1 is up-to-date
tests_redis_monitor_1 is up-to-date
tests_nginx_1 is up-to-date
tests_runner_1 is up-to-date
> docker exec -t tests_runner_1 sh -c "TARGET=php56 vendor/bin/phpunit"
stty: standard input
PHPUnit 6.2.3 by Sebastian Bergmann and contributors.

................           16 / 16 (100%)

Time: 1.39 seconds, Memory: 4.00MB

OK (16 tests, 54 assertions)
> docker exec -t tests_runner_1 sh -c "TARGET=php70 vendor/bin/phpunit"
stty: standard input
PHPUnit 6.2.3 by Sebastian Bergmann and contributors.

................           16 / 16 (100%)

Time: 1.29 seconds, Memory: 4.00MB

OK (16 tests, 54 assertions)
> docker exec -t tests_runner_1 sh -c "TARGET=php71 vendor/bin/phpunit"
stty: standard input
PHPUnit 6.2.3 by Sebastian Bergmann and contributors.

................           16 / 16 (100%)

Time: 1.08 seconds, Memory: 4.00MB

OK (16 tests, 54 assertions)
```


### Running the tests against the native phpredis handler

You can easily run the same test suite against the native phpredis handler.

To do so, comment out the line in `tests/webroot/visit-counter.php` where RedisSessionHandler is
enabled and the FPM container will automatically choose the phpredis save handler (version 3.1.2 at the time of writing).

```php
// session_set_save_handler(new \UMA\RedisSessionHandler(), true);
```

```bash
$ composer test
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Generating autoload files
> docker-compose -f tests/docker-compose.yml up -d
tests_redis_1 is up-to-date
tests_fpm56_1 is up-to-date
tests_fpm71_1 is up-to-date
tests_fpm70_1 is up-to-date
tests_redis_monitor_1 is up-to-date
tests_nginx_1 is up-to-date
tests_runner_1 is up-to-date
> docker exec -t tests_runner_1 sh -c "TARGET=php56 vendor/bin/phpunit"
stty: standard input
PHPUnit 6.2.3 by Sebastian Bergmann and contributors.

...FFF..FF......           16 / 16 (100%)

Time: 1.15 seconds, Memory: 4.00MB

There were 5 failures:

~~snip~~
```


### Manual testing

The `docker-compose.yml` file is configured to expose a random TCP port linked to the nginx container port 80. After running
`composer env-up` or `composer test` you can see which one was assigned with `docker ps`. With that knowledge you can
poke the testing webserver directly from your local machine using either a regular browser, cURL, wrk or similar tools.

Depending on your Docker setup you might need to replace `localhost` with the IP of the virtual machine that actually runs the Docker daemon:

```bash
$ curl -i localhost:32768/visit-counter.php?with_custom_cookie_params
HTTP/1.1 200 OK
Server: nginx/1.13.1
Date: Sat, 15 Jul 2017 09:40:25 GMT
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
Set-Cookie: PHPSESSID=tqm3akv0t5gdf25lf1gcspn479aa989jp0mltc3nuun2b47c7g10; expires=Sun, 16-Jul-2017 09:40:25 GMT; Max-Age=86400; path=/; secure; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache

1
```


CHANGELOG.md
### [0.9.9] - 2021-12-12

  * (Hotfix) Unlock the required version of `symfony/polyfill-php70`

### [0.9.8] - 2021-12-12

  * (Feature) Support all recent versions of PHP from 7.3 through 8.1 (contributed by [@wgirhad](https://github.com/wgirhad))

### [0.9.7] - 2020-04-07

  * (Improvement) Implemented support for Unix paths as means to connect to a Redis server.
  * (Bugfix) Ensure RedisSessionHandler::read() always returns a string (contributed by [therosco](https://github.com/therosco)).
  * (Docs) Acknowledged `redis.session.locking_enabled` INI directive from the native extension (starting from v4.1.0).
  * (Improvement) Updated versions of testing containers.

### [0.9.6] - 2018-03-31

  * (Bugfix) Fixed regenerated sessions expiration date (contributed by [@kavacky](https://github.com/kavacky))
  * (Bugfix) Fixed a phpredis 4.0 BC break in the return type of the `exists()` function.
  * (Improvement) PHP 7.2 was released since 0.9.5, and a 7.2 FPM daemon has been added to the test pool.

### [0.9.5] - 2017-07-13

  * (Bugfix) Made the handler acknowledge custom session cookie parameters set with `session_set_cookie_params` (contributed by [@scottlucas](https://github.com/scottlucas))
  * (Improvement) Expanded integration tests to cover all supported major versions of PHP (5.6, 7.0 and 7.1).

### [0.9.4] - 2017-07-10

  * (Bugfix) Prevented the handler from hanging when the `max_execution_time` directive is set to `0` (contributed by [@scottlucas](https://github.com/scottlucas))

### [0.9.3] - 2017-04-28

  * (Bugfix) Made the handler compatible with PHP +7.1.2 by avoiding the call to `session_regenerate_id` inside the handler.

### [0.9.2] - 2017-02-20

  * (Improvement) Set up continuous integration based on Scrutinizer.
  * (Feature) Added support for `timeout`, `prefix`, `auth` and `database` query params on the `session.save_path` directive, as in the native handler.

### [0.9.1] - 2017-02-05

  * (Improvement) Introduced exponential backoff between session locking attempts.
  * (Improvement) A session ID is now never locked before checking that it doesn't need to be regenerated.

### [0.9.0] - 2017-02-05

  * Initial pre-release

[0.9.7]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.8...v0.9.9
[0.9.7]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.7...v0.9.8
[0.9.7]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.6...v0.9.7
[0.9.6]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.5...v0.9.6
[0.9.5]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.4...v0.9.5
[0.9.4]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.3...v0.9.4
[0.9.3]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.2...v0.9.3
[0.9.2]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.1...v0.9.2
[0.9.1]: https://github.com/1ma/RedisSessionHandler/compare/v0.9.0...v0.9.1
[0.9.0]: https://github.com/1ma/RedisSessionHandler/tree/b6b149a3d5322e49a3c4c933ed8154ad3da30570


tests/e2e/BasicTest.php
<?php

namespace UMA\RedisSessions\Tests\E2E;

use GuzzleHttp\Psr7\Request;
use Psr\Http\Message\ResponseInterface;
use UMA\SavePathParser;

class BasicTest extends EndToEndTestCase
{
    /**
     * This test sends an anonymous request to a PHP
     * script that will create a session.
     *
     * After this, the response should have a 'Set-Cookie' header
     * with the new session ID, its body should be '1', and Redis
     * should have exactly one entry.
     */
    public function testAnonymousRequest()
    {
        $response = $this->http->send(
            new Request('GET', '/visit-counter.php')
        );

        $this->assertSame('1', (string) $response->getBody());

        $this->assertCreatedNewSession($response);

        $this->assertSame(1, $this->redis->dbSize());
    }

    /**
     * This test sends two anonymous, unrelated requests to a PHP
     * script that will create a session for each one of them.
     *
     * After this, each response should have a Set-Cookie header
     * with its new session ID (each one should be different), and
     * both their bodies must be '1'. Redis should have exactly
     * two entries.
     */
    public function testUnrelatedAnonymousRequests()
    {
        $firstResponse = $this->http->send(
            new Request('GET', '/visit-counter.php')
        );

        $secondResponse = $this->http->send(
            new Request('GET', '/visit-counter.php')
        );

        $this->assertSame('1', (string) $firstResponse->getBody());
        $this->assertSame('1', (string) $secondResponse->getBody());

        $this->assertCreatedNewSession($firstResponse);
        $this->assertCreatedNewSession($secondResponse);
        $this->assertNotSame($firstResponse->getHeaderLine('Set-Cookie'), $secondResponse->getHeaderLine('Set-Cookie'));

        $this->assertSame(2, $this->redis->dbSize());
    }

    /**
     * This test sends an anonymous request, then a second one
     * authenticated with the cookie received in the first.
     *
     * After this, the first response should have a 'Set-Cookie' header
     * with its new session ID, and its body should be '1'.The second
     * response should NOT have a 'Set-Cookie' header and its body
     * should be '2'. Redis should have exactly one entry.
     */
    public function testRelatedRequests()
    {
        $firstResponse = $this->http->send(
            new Request('GET', '/visit-counter.php')
        );

        $secondResponse = $this->http->send(
            new Request('GET', '/visit-counter.php', $this->prepareSessionHeader($firstResponse))
        );

        $this->assertSame('1', (string) $firstResponse->getBody());
        $this->assertSame('2', (string) $secondResponse->getBody());

        $this->assertCreatedNewSession($firstResponse);
        $this->assertFalse($secondResponse->hasHeader('Set-Cookie'));

        $this->assertSame(1, $this->redis->dbSize());
    }

    /**
     * This test sends a malicious request attempting a session
     * fixation attack.
     *
     * After this, the response should have a 'Set-Cookie' header with
     * a newly generated ID and its body should be '1'. Redis should
     * have exactly one entry, and its key should not be 'madeupkey'.
     *
     * Additionally, the Set-Cookie header MUST NOT have an 'expires' tag. This
     * was a bug where a forcefully regenerated session ID used to be stamped
     * with an instant expiration time.
     */
    public function testMaliciousRequest()
    {
        $response = $this->http->send(
            new Request('GET', '/visit-counter.php', ['Cookie' => 'PHPSESSID=madeupkey;'])
        );

        $this->assertSame('1', (string) $response->getBody());

        $this->assertCreatedNewSession($response);
        $this->assertNotRegExp('/expires=/i', $response->getHeaderLine('Set-Cookie'));

        $this->assertSame(1, $this->redis->dbSize());
        $this->assertFalse($this->redis->get(SavePathParser::DEFAULT_PREFIX.'madeupkey'));
    }

    /**
     * This test sends a malicious request attempting a session
     * fixation attack but uses custom cookie params.
     *
     * After this, the response should have a 'Set-Cookie' header with
     * a newly generated ID and its body should be '1'. The cookie should
     * have custom parameters as per visit-counter.php and Redis should
     * have exactly one entry, and its key should not be 'madeupkey'.
     */
    public function testMaliciousRequestWithCustomCookieParams()
    {
        $response = $this->http->send(
            new Request('GET', '/visit-counter.php?with_custom_cookie_params', ['Cookie' => 'PHPSESSID=madeupkey;'])
        );

        $this->assertSame('1', (string) $response->getBody());

        $this->assertCustomCookieParams($response);

        $this->assertSame(1, $this->redis->dbSize());
        $this->assertFalse($this->redis->get(SavePathParser::DEFAULT_PREFIX.'madeupkey'));
    }


    /**
     * This test checks that the server behaves correctly when a previously valid
     * session ID is removed from Redis in between requests.
     *
     * @see https://github.com/1ma/RedisSessionHandler/issues/3
     */
    public function testFlushedDatabase()
    {
        $firstResponse = $this->http->send(
            new Request('GET', '/visit-counter.php')
        );

        $this->assertSame('1', (string) $firstResponse->getBody());
        $this->assertCreatedNewSession($firstResponse);

        $this->redis->flushAll();

        $this->assertSame(0, $this->redis->dbSize());

        $secondResponse = $this->http->send(
            new Request('GET', '/visit-counter.php', $this->prepareSessionHeader($firstResponse))
        );

        $this->assertSame('1', (string) $secondResponse->getBody());
        $this->assertCreatedNewSession($secondResponse);

        $this->assertSame(1, $this->redis->dbSize());
    }

    /**
     * This test checks that the server behaves correctly when a script is run that has no time limit
     */
    public function testUnlimitedExecutionTime()
    {
        $response = $this->http->send(
            new Request('GET', '/visit-counter.php?with_no_time_limit')
        );

        $this->assertSame('1', (string) $response->getBody());
        $this->assertCreatedNewSession($response);
        $this->assertSame(1, $this->redis->dbSize());
    }

    /**
     * Asserts whether a received request triggered the creation of a new session.
     * It does so by searching for and inspecting the 'Set-Cookie' header.
     *
     * @param ResponseInterface $response
     */
    protected function assertCreatedNewSession(ResponseInterface $response)
    {
        $this->assertTrue($response->hasHeader('Set-Cookie'));
        $this->assertStringStartsWith('PHPSESSID=', $response->getHeaderLine('Set-Cookie'));
    }

    /**
     * Asserts whether a received request triggered the creation of a new session with example custom parameters.
     * It does so by searching for and inspecting the 'Set-Cookie' header.
     * 
     * Expected result:
     * set-cookie:PHPSESSID=0691f472784601a324eadf78c05cf4c1; expires=Tue, 11-Jul-2017 14:58:34 GMT; Max-Age=86400; path=/; secure; HttpOnly
     *
     * @param ResponseInterface $response
     */
    protected function assertCustomCookieParams(ResponseInterface $response)
    {
        $this->assertTrue($response->hasHeader('Set-Cookie'));
        $this->assertStringStartsWith('PHPSESSID=', $response->getHeaderLine('Set-Cookie'));
        $this->assertRegExp('/expires=/i', $response->getHeaderLine('Set-Cookie'));
        $this->assertRegExp('/path=/i', $response->getHeaderLine('Set-Cookie'));
        $this->assertRegExp('/secure/i', $response->getHeaderLine('Set-Cookie'));
        $this->assertRegExp('/HttpOnly/i', $response->getHeaderLine('Set-Cookie'));
    }
}





tests/e2e/ConcurrentTest.php
<?php

namespace UMA\RedisSessions\Tests\E2E;

use GuzzleHttp\Pool;
use GuzzleHttp\Psr7\Request;

class ConcurrentTest extends EndToEndTestCase
{
    const CONCURRENCY_LEVEL = 20;
    const REQUESTS_PER_TEST = 200;

    /**
     * This test sends a barrage of anonymous requests.
     *
     * After this, Redis should contain exactly REQUESTS_PER_TEST entries.
     */
    public function testAnonymousRequests()
    {
        $this->massSend(
            new Request('GET', '/visit-counter.php')
        );

        $this->assertSame(self::REQUESTS_PER_TEST, $this->redis->dbSize());
    }

    /**
     * This test first sends an anonymous request to trigger the creation of
     * a new session, then a barrage of authenticated requests, then a last request.
     *
     * After this, Redis should contain exactly one entry, and it's number of
     * visits should be exactly REQUESTS_PER_TEST + 2
     *
     * This is the crucial test that exercises session locking. You can temporarily
     * comment the spinlock in the RedisSessionHandler class to induce concurrency
     * bugs and see the test fail.
     */
    public function testAuthenticatedRequests()
    {
        $firstResponse = $this->http->send(
            new Request('GET', '/visit-counter.php')
        );

        $this->massSend(
            new Request('GET', '/visit-counter.php', $this->prepareSessionHeader($firstResponse))
        );

        $lastResponse = $this->http->send(
            new Request('GET', '/visit-counter.php', $this->prepareSessionHeader($firstResponse))
        );

        $this->assertSame((string)(self::REQUESTS_PER_TEST + 2), (string) $lastResponse->getBody());

        $this->assertSame(1, $this->redis->dbSize());
    }

    /**
     * This test sends a barrage of requests attempting a session fixation attack.
     *
     * After this, the cache should contain exactly REQUESTS_PER_TEST entries.
     * Finding just one entry in Redis would mean that the attack was successful.
     */
    public function testMaliciousRequests()
    {
        $this->massSend(
            new Request('GET', '/visit-counter.php', ['Cookie' => 'PHPSESSID=madeupkey;'])
        );

        $this->assertSame(self::REQUESTS_PER_TEST, $this->redis->dbSize());
    }

    /**
     * A helper that asynchronously sends multiple identical requests to the web server,
     * then waits for all those requests to complete.
     *
     * @param Request $request
     */
    protected function massSend(Request $request)
    {
        $replicator = function (Request $request) {
            for ($i = 0; $i < self::REQUESTS_PER_TEST; ++$i) {
                yield clone $request;
            }
        };

        (new Pool(
            $this->http,
            $replicator($request->withProtocolVersion('1.0')),
            ['concurrency' => self::CONCURRENCY_LEVEL]
        ))->promise()->wait();
    }
}


tests/e2e/EndToEndTestCase.php
<?php

namespace UMA\RedisSessions\Tests\E2E;

use GuzzleHttp\Client;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\ResponseInterface;
use Redis;

abstract class EndToEndTestCase extends TestCase
{
    /**
     * @var Client
     */
    protected $http;

    /**
     * @var Redis
     */
    protected $redis;

    /**
     * e2e test cases start creating an HTTP client pointing
     * to the testing web server and flushing the Redis database.
     */
    public function setUp()
    {
        $this->http = new Client(['base_uri' => 'http://testapp']);

        $this->redis = new Redis();
        $this->redis->connect('redis');
        $this->redis->flushAll();
    }

    /**
     * A helper that takes a Guzzle response with a 'Set-Cookie'
     * header and prepares the headers array for a subsequent
     * authenticated request.
     *
     * @param ResponseInterface $response
     *
     * @return array
     *
     * @example ['Cookie' => 'PHPSESSID=urbigl47u82h0r9qke0q8jt23836j4gos5kebb0ed5ukiqepsau0;']
     */
    protected function prepareSessionHeader(ResponseInterface $response)
    {
        return [
            'Cookie' => explode(' ', $response->getHeaderLine('Set-Cookie'))[0]
        ];
    }
}


src/SavePathParser.php
<?php

namespace UMA;

/**
 * @author Marcel Hernandez
 */
class SavePathParser
{
    /**
     * Redis port to use when it is not specified in the save path
     */
    const DEFAULT_PORT = 6379;

    /**
     * Connection timeout to use when it is not specified in the save path
     */
    const DEFAULT_TIMEOUT = 0.0;

    /**
     * Session prefix to use when it is not specified in the save path
     */
    const DEFAULT_PREFIX = 'PHPREDIS_SESSION:';

    /**
     * Authentication string to use when it is not specified in the save path
     */
    const DEFAULT_AUTH = null;

    /**
     * Redis database to use when it is not specified in the save path
     */
    const DEFAULT_DATABASE = 0;

    const DEFAULT_OPTIONS = [
        'timeout' => self::DEFAULT_TIMEOUT,
        'prefix' => self::DEFAULT_PREFIX,
        'auth' => self::DEFAULT_AUTH,
        'database' => self::DEFAULT_DATABASE
    ];

    /**
     * @param string $path The session.save_path parameter from php.ini
     *
     * @return array [host, port, connection timeout, session prefix, auth, database]
     *
     * @example 'localhost' => ['localhost', 6379, 0.0, 'PHPREDIS_SESSION:', null, 0]
     * @example 'tcp://1.2.3.4:5678?auth=secret&database=3' => ['1.2.3.4', 5678, 0.0, 'PHPREDIS_SESSION:', 'secret', 3]
     */
    public static function parse($path)
    {
        // parse_url() is intended to parse URLs (but not URIs) so it cannot parse Unix
        // socket paths. However for backwards compatibility reasons it can still parse
        // URIs that start with 'file://'
        //
        // @see https://www.php.net/manual/en/function.parse-url.php
        if (0 === strpos($path, 'unix://')) {
            $path = str_replace('unix://', 'file://', $path);
        }

        $parsed_path = parse_url($path);

        $host = isset($parsed_path['host']) ?
            $parsed_path['host'] : $parsed_path['path'];

        $port = isset($parsed_path['port']) ?
            $parsed_path['port'] : static::DEFAULT_PORT;

        $opts = [];

        if (isset($parsed_path['query'])) {
            parse_str($parsed_path['query'], $opts);
        }

        $opts = array_merge(static::DEFAULT_OPTIONS, $opts);

        return [
            $host, $port, (float) $opts['timeout'], $opts['prefix'], $opts['auth'], (int) $opts['database']
        ];
    }
}
